src/NuGet.Services.EndToEnd/Support/PushedPackagesFixture.cs

```diff
@@ -11,108 +11,178 @@
         {
-            PushedPackage pushedPackage;
-            lock (_packagesLock)
+            var pushedPackage = GetPushedPackage(requestedPackageType, logger);
+            if (pushedPackage != null)
             {
-                if (!_packages.TryGetValue(packageType, out pushedPackage))
-                {
-                    var package = CreatePackage(packageType);
-                    pushedPackage = new PushedPackage(package);
-                    _packages[packageType] = pushedPackage;
-                }
+                return pushedPackage;
             }
 
-            if (pushedPackage.Pushed)
-            {
-                logger.WriteLine($"Package {pushedPackage} push has already been pushed.");
-                return pushedPackage.Package;
-            }
+            return await PushUnpushedPackagesAsync(requestedPackageType, logger);
+        }
 
-            var acquired = await pushedPackage.PushLock.WaitAsync(0);
+        private async Task<Package> PushUnpushedPackagesAsync(PackageType requestedPackageType, ITestOutputHelper logger)
+        {
+            var acquired = await _pushLock.WaitAsync(0);
             try
             {
                 if (!acquired)
                 {
-                    logger.WriteLine($"Another test is already pushing {pushedPackage}. Waiting.");
-                    await pushedPackage.PushLock.WaitAsync();
-                    acquired = true;
+                    logger.WriteLine("Another test is already pushing the packages. Waiting.");
+                    await _pushLock.WaitAsync();
+                }
+
+                // Has the package already been pushed?
+                var pushedPackage = GetPushedPackage(requestedPackageType, logger);
+                if (pushedPackage != null)
+                {
+                    return pushedPackage;
                 }
 
-                if (pushedPackage.Pushed)
+                var packageTypes = GetPackageTypes(requestedPackageType);
+
+                // Push all of the package types that have not been pushed yet.
+                var pushTasks = new Dictionary<PackageType, Task<Package>>();
+                foreach (var packageType in packageTypes)
                 {
-                    logger.WriteLine($"Package {pushedPackage} has already been pushed.");
-                    return pushedPackage.Package;
+                    if (!_packages.TryGetValue(packageType, out Package package))
+                    {
+                        pushTasks.Add(packageType, UncachedPushAsync(requestedPackageType, packageType, logger));
+                    }
                 }
 
-                using (var packageStream = new MemoryStream(pushedPackage.Package.NupkgBytes.ToArray()))
+                await Task.WhenAll(pushTasks.Values);
+
+                // Add the pushed packages to the cache.
+                lock (_packagesLock)
```

Comment:
You could remove the code inside this `lock` if you added `_packages[requestedPackageType] = package` to `UncachedPushAsync`.